.github/workflows/build.yaml

```diff
@@ -253,6 +254,11 @@ jobs:
       - pre-build-cli
       - pre-build-go-binaries
       - pre-build-docs
+    # This step will run even if required pre-* steps failed. By following this
+    # approach as much of the build matrix as possible is completed despite
+    # transient issues e.g. docker pull timeouts. In this way some e2e jobs that
+    # require some part of the build matrix can still proceed.
+    if: ${{ !cancelled() }}
```

Comment:
It is unfortunate that `if` does not respect matrix (according to a note 
[here](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idif)).

I'm a bit concerned that now an error will cascade through the workflow causing these later stages break with weird errors just because the failing earlier stage did not create something somewhere.

Even worse if later stages will succeed despite some things missing earlier.

There seems to be a way to have a matrix and group it too, via subworkflows: <https://github.com/orgs/community/discussions/42335#discussioncomment-7997348>. That'll be a bit of refactoring to our pipeline but we're far away from a state we once were, when everything was encapsulated in shell scripts that could work in any CI. With our current state of GHA lock-in, maybe that's fine, just more work on us?

Overall, I'm also ok with this current approach in this PR and would not block merging it if @janisz is strongly for (which I suspect could be the case).

Reply:
I see your point @msugakov but based on testing I think the failures are safely propagated with missing inputs in the subsequent steps. Though I see that there is a bug in the download artifact with retry action as it does not fail when the UI artifact is missing: https://github.com/stackrox/stackrox/actions/runs/9573456734/job/26395165226#step:10:25. I will address that in another PR.

Reply:
download-artifact-with-retry also did not retry: https://github.com/stackrox/stackrox/pull/11637